File: app/src/main/java/com/example/mvvmlib/adapter/BaseRecyclerAdapter.java
package com.example.mvvmlib.adapter;

import android.content.Context;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.databinding.ObservableArrayList;
import androidx.recyclerview.widget.RecyclerView;

import org.jetbrains.annotations.NotNull;

import java.util.List;

public abstract class BaseRecyclerAdapter<T> extends RecyclerView.Adapter<AbsViewHolder> {
    protected Context mContext;
    private LayoutInflater mInflater;
    protected ObservableArrayList<T> mDataList;
    private OnItemClickListener mClickListener;

    public BaseRecyclerAdapter(Context context, ObservableArrayList<T> data) {
        this.mContext = context;
        mInflater = LayoutInflater.from(mContext);
        mDataList = data;
    }

    public void setOnItemClickListener(OnItemClickListener mClickListener) {
        this.mClickListener = mClickListener;
    }

    @Override
    public AbsViewHolder<T> onCreateViewHolder(@NonNull ViewGroup viewGroup, int viewType) {
        return onFindViewHolder(viewGroup, viewType);
    }

    public abstract @NotNull
    AbsViewHolder onFindViewHolder(@NonNull ViewGroup viewGroup, int viewType);

    @Override
    public void onBindViewHolder(@NonNull AbsViewHolder absViewHolder, int position) {
        if (mDataList == null || position >= mDataList.size()) return;
        T data = mDataList.get(position);
        if (data != null) {
            absViewHolder.onBindViewHolder(position, mDataList.size(), data);
        }
    }


    @Override
    public int getItemCount() {
        return mDataList == null ? 0 : mDataList.size();
    }

    public List<T> getData() {
        return mDataList;
    }

    /***
     * 添加数据源
     * @param data
     */
    public void addItemData(T data) {
        if (mDataList == null) {
            mDataList = new ObservableArrayList<>();
        }
        if (data != null && !mDataList.contains(data)) {
            mDataList.add(data);
        }
    }

    /***
     * 添加一组数据
     * @param data
     */
    public void addAllData(List<? extends T> data) {
        if (data == null) return;
        if (mDataList == null) {
            mDataList = new ObservableArrayList<>();
        }
        mDataList.addAll(data);
    }

    /***
     * 设置单条数据
     * @param postion
     * @param data
     */
    public void setItemData(int postion, T data) {
        if (mDataList == null || postion > mDataList.size()) return;
        mDataList.set(postion, data);
    }

    /***
     * 清空所有数据
     */
    public void clearData() {
        if (mDataList != null && !mDataList.isEmpty()) {
            mDataList.clear();
        }
    }

    /***
     * 更新数据集
     * @param data
     */
    public void setData(ObservableArrayList data) {
        this.mDataList = data;
    }

    /***
     * 加载布局
     * @param resLayout
     * @param parent
     * @return
     */
    public View inflaterView(int resLayout, ViewGroup parent) {
        if (mInflater == null) {
            mInflater = LayoutInflater.from(mContext);
        }
        return mInflater.inflate(resLayout, parent, false);
    }

    public interface OnItemClickListener<T> {
        void onItemClick(RecyclerView.ViewHolder holder, int position, T data);
    }

}

File: bindingadapter/src/main/java/com/example/bindingadapter/base/BaseRecyclerAdapter.kt
package com.example.bindingadapter.base

import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.annotation.LayoutRes
import androidx.databinding.ObservableArrayList
import androidx.recyclerview.widget.RecyclerView
import com.example.bindingadapter.BR
import com.example.bindingadapter.R
import com.example.bindingadapter.viewholder.AbsViewHolder
import com.example.bindingadapter.viewholder.NullTypeViewHolder

abstract class BaseRecyclerAdapter<T> :
    RecyclerView.Adapter<AbsViewHolder<T, *>>() {
    protected val mData = ObservableArrayList<T>()
    private var mInflater: LayoutInflater? = null
    var onItemClickListener: OnItemClickListener<T>? = null

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): AbsViewHolder<T, *> {
        var holder = onFindViewHolder(parent, viewType)
        return holder ?: NullTypeViewHolder(inflaterView(R.layout.holder_null_type_holder, parent))
    }

    override fun getItemCount(): Int {
        return mData.size
    }

    override fun onBindViewHolder(holder: AbsViewHolder<T, *>, position: Int) {
        holder.bindData(mData[position])
        if (holder !is NullTypeViewHolder && onItemClickListener != null) {
            holder.itemBinding?.setVariable(BR.onClickListener, object : View.OnClickListener {
                override fun onClick(v: View?) {
                    onItemClickListener?.onItemClick(
                        holder, holder.getLayoutPosition(),
                        mData[holder.getLayoutPosition()]
                    )
                }

            })
        }
    }

//    override fun onViewAttachedToWindow(holder: AbsViewHolder<T, *>) {
//        super.onViewAttachedToWindow(holder)
//        holder.bind()
//    }
//
//    override fun onViewDetachedFromWindow(holder: AbsViewHolder<T, *>) {
//        super.onViewDetachedFromWindow(holder)
//        holder.unBind()
//    }

    fun inflaterView(@LayoutRes resLayout: Int, parent: ViewGroup): View {
        if (mInflater == null) {
            mInflater = LayoutInflater.from(parent.context)
        }
        return mInflater!!.inflate(resLayout, parent, false)
    }

    /***
     * 添加单一数据
     */
    fun addItemData(data: T) {
        if (data != null && !mData.contains(data)) {
            mData.add(data)
        }
    }

    /***
     * 添加一组数据
     */
    fun addAllData(data: List<T>) {
        if (data.isNullOrEmpty()) return
        mData.addAll(data)
    }

    /***
     * 改变单一数据
     */
    fun setItemData(position: Int, data: T) {
        if (position < 0 || position > mData.size) return
        mData[position] = data
    }

    /***
     * 清空数据
     */
    fun clearData() {
        if (mData.isNotEmpty()) {
            mData.clear()
        }
    }

    /***
     * 替换所有数据
     */
    fun setData(data: List<T>) {
        clearData()
        addAllData(data)
    }

    abstract fun onFindViewHolder(viewGroup: ViewGroup, viewType: Int): AbsViewHolder<T, *>?

}


interface OnItemClickListener<T> {
    fun onItemClick(holder: RecyclerView.ViewHolder, position: Int, data: T)
}
